### PR TITLE
6281 create customer for transaction

### PR DESCRIPTION
**Pull request by X0nic**
This is a continuation of #317 use this link to see the differences from that PR. https://github.com/sassafrastech/madeline/compare/6280_add_disbursment_date...6281_create_customer_for_transaction

### CODE REVIEW

File: app/models/accounting/quickbooks/transaction_creator.rb [hunk 1]
```diff
@@ -67,9 +72,11 @@ def create_line_item(amount:, loan_id:, posting_type:, description:, qb_account_
         line_item
       end
 
-      # We are not creating a customer here, but a customer reference.
+      # We are may be creating a customer here if needed. We return the qb_id and manualy create a reference.
```
Comment (hooverlunch) [inline app/models/accounting/quickbooks/transaction_creator.rb]: Grammar

File: app/models/accounting/quickbooks/transaction_creator.rb [hunk 2]
```diff
@@ -78,6 +85,17 @@ def create_customer_reference(qb_customer_id:)
...
+        return organization.qb_id if organization.qb_id.present?
+
+        qb_customer = ::Quickbooks::Model::Customer.new
+        qb_customer.display_name = organization.name
+
+        new_qb_customer = customer_service.create(qb_customer)
+
+        new_qb_customer.id
+      end
+
```
Comment (hooverlunch) [inline app/models/accounting/quickbooks/transaction_creator.rb]: This is feeling a bit bloat-y to me, i.e. violation of SRP. Can we make a new class `Customer' that is initialized with an Organization and has public method `reference`, which calls private method `find_or_create...`?

Then anywhere you need a customer reference you do `Customer.new(organization).reference`?

... and don't forget the class description comment :)

Comment (hooverlunch) [inline app/models/accounting/quickbooks/transaction_creator.rb]: And in general, please try to be careful about adding too much responsibility to a class like this. If you're creating and managing customers inside something called TransactionCreator, that is a hint that you need a new class.

File: spec/models/accounting/quickbooks/transaction_creator_spec.rb [hunk 3]
```diff
@@ -2,28 +2,109 @@
...
+    context 'and date is supplied' do
+      let(:date) { 3.days.ago }
+
+      it 'creates JournalEntry with date' do
+        expect(generic_service).to receive(:create) do |arg|
+          expect(arg.txn_date).to eq date
+        end
+        subject
+      end
+    end
```
Comment (hooverlunch) [inline spec/models/accounting/quickbooks/transaction_creator_spec.rb]: The above two contexts should move into a new Customer spec. Then this spec can just stub the customer fetching stuff and assume it works properly. :)

File: spec/models/accounting/quickbooks/customer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  context 'when qb customer does exist' do
+    let(:organization) { create(:organization, qb_id: qb_customer_id) }
+    let(:service) { instance_double(Quickbooks::Service::Customer, create: nil) }
+
+    it 'does not create a customer' do
+      expect(service).not_to receive(:create)
+      subject.reference
+    end
+
+    it 'returns proper customer reference' do
```
Comment (X0nic) [inline spec/models/accounting/quickbooks/customer_spec.rb]: Seems like overkill to extract out a shared example for two instances on one test. Any suggestions?

File: app/models/accounting/quickbooks/customer.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        qb_customer = ::Quickbooks::Model::Customer.new
+        qb_customer.display_name = organization.name
+
+        new_qb_customer = service.create(qb_customer)
+
+        new_qb_customer.id
+      end
+    end
+  end
+end
```
Comment (hooverlunch) [inline app/models/accounting/quickbooks/customer.rb]: Looks great! So much nicer to have this in its own place. Thanks!

**Comment by hooverlunch** [comment]
Also thanks for the compare link. Very thoughtful!

File: spec/models/accounting/quickbooks/transaction_creator_spec.rb [hunk 6]
```diff
@@ -2,28 +2,89 @@
```
Comment (emitche) [inline spec/models/accounting/quickbooks/transaction_creator_spec.rb]: @X0nic I've made a guess about how best to merge the earlier develop (which includes changes from 6280) with this file. If you look over develop once the release goes out and something isn't right, just let me know.

**Pull request merged**